src/framebuffer/swtfb_client.rs

```diff
@@ -0,0 +1,224 @@
+    pub y2: i32,
+
+    pub waveform: i32,
+    pub flags: i32,
+}
+
+#[derive(Debug, Clone, Copy)]
+#[allow(non_camel_case_types, dead_code)]
+pub struct wait_sem_data {
+    /// C string
+    pub sem_name: [u8; 512],
+}
+
+/// MSG_TYPE has to match swtfb_update_data !!!
+#[derive(Clone, Copy)]
+#[allow(non_camel_case_types, dead_code)]
+pub struct swtfb_update {
+    pub mtype: MSG_TYPE,
+    pub data: swtfb_update_data,
+    //ms: u64,
+}
+
+#[derive(Clone, Copy)]
+#[allow(non_camel_case_types, dead_code)]
+pub union swtfb_update_data {
+    pub xochitl_update: xochitl_data,
+    pub update: mxcfb_update_data,
+    pub wait_update: wait_sem_data,
+}
+
+pub struct SwtfbClient {
+    msqid: i32,
+    do_wait_ioctl: bool,
+}
+
+impl Default for SwtfbClient {
+    fn default() -> Self {
+        let msqid = unsafe {
+            libc::msgget(
+                SWTFB_MESSAGE_QUEUE_ID,
+                libc::IPC_CREAT | libc::SHM_R | libc::SHM_W,
+            )
+        };
+        assert!(msqid >= 0);
+
+        // Does nested need special handling?
+        // This may not be needed at all.
+        if env::var("RM2FB_ACTIVE").is_ok() {
+            env::set_var("RM2FB_NESTED", "1");
+        } else {
+            env::set_var("RM2FB_ACTIVE", "1");
+        }
+
+        Self {
+            msqid,
+            do_wait_ioctl: env::var("RM2FB_NO_WAIT_IOCTL").is_err(),
+        }
+    }
+}
+
+impl SwtfbClient {
+    pub fn open_buffer(&self) -> Result<(File, MmapRaw), IoError> {
+        let device = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open("/dev/shm/swtfb.01")?;
+        let ret = unsafe { libc::ftruncate(device.as_raw_fd(), BUF_SIZE as libc::off_t) };
+        if ret < 0 {
+            return Err(IoError::last_os_error());
+        }
+        let mem_map = MmapOptions::new().len(BUF_SIZE as usize).map_raw(&device)?;
+        Ok((device, mem_map))
+    }
+
+    pub fn send(&self, update: &swtfb_update) -> bool {
+        unsafe {
+            let ptr = ptr::addr_of!(*update) as *const c_void;
+            libc::msgsnd(self.msqid, ptr, mem::size_of::<swtfb_update>(), 0) == 0
+        }
+    }
+
+    pub fn send_mxcfb_update(&self, update: &mxcfb_update_data) -> bool {
+        self.send(&swtfb_update {
+            mtype: MSG_TYPE::UPDATE_t,
+            data: swtfb_update_data { update: *update },
+        })
+    }
+
+    pub fn send_xochitl_update(&self, data: &xochitl_data) -> bool {
+        self.send(&swtfb_update {
+            mtype: MSG_TYPE::XO_t,
+            data: swtfb_update_data {
+                xochitl_update: *data,
+            },
+        })
+    }
+
+    pub fn wait_for_update_complete(&self) {
+        if !self.do_wait_ioctl {
+            return;
+        }
+
+        // https://github.com/ddvk/remarkable2-framebuffer/blob/1e288aa9/src/client/main.cpp#L149
+
+        let sem_name_str = format!("/rm2fb.wait.{}", unsafe { libc::getpid() });
+        let mut sem_name = [0u8; 512];
```

Comment:
```suggestion
        let mut sem_name = [0u8; 12 + 10];
```
`pid_t` is `i32` => 10 digits tops